File: Data Structures & Algorithms/binary-tree-diameter/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode *root, int &dia) {
        if(!root) return 0;

        int lh = dfs(root->left, dia);
        int rh = dfs(root->right, dia);

        dia = max(lh + rh, dia);
        return (max(lh, rh) + 1);
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int dia = 0;
        dfs(root, dia);
        return dia;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root: TreeNode | null): number {
        return this.#dfs(root, -1e5);
    }

    #dfs(root: TreeNode | null, high: number): number {
        if(!root) return 0;
        const cur = +(root.val >= high);
        high = Math.max(high, root.val);
        return cur + this.#dfs(root.left, high) + this.#dfs(root.right, high);
    }
}
